synchronization/supervised/shutdown.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2019 the orbs-network-go authors
+// This file is part of the orbs-network-go library in the Orbs project.
+//
+// This source code is licensed under the MIT license found in the LICENSE file in the root directory of this source tree.
+// The above notice should be included in all copies or substantial portions of the software.
+
+package supervised
+
+import (
+	"context"
+	"fmt"
+	"github.com/orbs-network/scribe/log"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+)
+
+type ShutdownWaiter interface {
+	WaitUntilShutdown(shutdownContext context.Context)
+}
+
+type GracefulShutdowner interface {
+	ShutdownWaiter
+	GracefulShutdown(shutdownContext context.Context)
+}
+
+type ChanWaiter struct {
+	Closed      chan struct{}
+	description string
+}
+
+func (c *ChanWaiter) WaitUntilShutdown(shutdownContext context.Context) {
+	select {
+	case <-c.Closed:
+	case <-shutdownContext.Done():
```

Comment:
add if `shutownContext.Err() == timeout` to distinguish cancellation from shutdown in this case